eventually.go

```diff
@@ -0,0 +1,159 @@
+// It also allows providing a stable retry strategy to run a stability check,
+// i.e. once the check is successful, it keeps checking that it stays that way.
+type EventuallyChecker struct {
+	checker             Checker
+	retryStrategy       *retry.Strategy
+	stableRetryStrategy *retry.Strategy
+}
+
+// WithStrategy allows specifying a custom retry strategy, specifying initial
+// delay, delay between attempts, maximum duration before timing out, etc.
+func (e *EventuallyChecker) WithStrategy(strategy *retry.Strategy) *EventuallyChecker {
+	e.retryStrategy = strategy
+	return e
+}
+
+// WithStableStrategy allows specifying a custom retry strategy for the
+// stability check. If not provided, no stability check will be run.
+func (e *EventuallyChecker) WithStableStrategy(strategy *retry.Strategy) *EventuallyChecker {
+	e.stableRetryStrategy = strategy
+	return e
+}
+
+// Check implements Checker.Check by calling the given got function repeatedly
+// and calling the underlying Checker with the returned value, according to the
+// retry Strategy, until the verification succeeds or the Strategy times out.
+// If the Strategy times out, the Check will fail.
+//
+// If an additional stable retry Strategy is provided, it also calls the
+// underlying Checker again after a succesfull check, according to the stable
+// retry Strategy, until either the Strategy times out or the verification
+// fails. If the stable Strategy times out, the Check will succeed.
+func (e *EventuallyChecker) Check(got interface{}, args []interface{}, note func(key string, value interface{})) error {
+	// Validate that the given got parameter is a function with no parameters
+	// and one return value.
+	f := reflect.ValueOf(got)
+	if f.Kind() != reflect.Func {
+		note("got", got)
+		return BadCheckf("first argument is not a function")
+	}
+	ftype := f.Type()
+	if ftype.NumIn() != 0 {
+		note("function", got)
+		return BadCheckf("cannot use a function receiving arguments")
+	}
+	if ftype.NumOut() != 1 {
+		note("function", got)
+		return BadCheckf("cannot use a function returning more than one value")
+	}
+
+	// Run the checker according to the retry strategy, succeeding on the first
+	// successful check.
+	for i := e.retryStrategy.Start(); ; {
+		got = f.Call(nil)[0].Interface()
+		err := e.checker.Check(got, args, note)
+		if err == nil {
+			break
+		}
+		if !i.Next(nil) {
+			note("got", got)
+			return fmt.Errorf("tried for %v, %s", e.retryStrategy.MaxDuration, err.Error())
+		}
+	}
+
+	// If a stable strategy is provided, run the checker again according to it,
+	// failing on the first error.
+	if e.stableRetryStrategy != nil {
+		for i := e.stableRetryStrategy.Start(); ; {
+			got = f.Call(nil)[0].Interface()
+			err := e.checker.Check(got, args, note)
+			if err != nil {
+				note("got", got)
+				return fmt.Errorf("less than %v after an initial success, %s", e.stableRetryStrategy.MaxDuration, err.Error())
+			}
+			if !i.Next(nil) {
```

Comment:
Not familiar with retry, but is this needed?

Reply:
This Next call is effectively doing the iteration, returning whether there will be any more step or not. The way we wrote the for loop is a bit weird, there's no termination condition, the break here achieves that - I can write it in a somewhat nicer way like:
`for i := Start(); hasNext; hasNext = i.Next(nil)` which is probably what I will do.